Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/ADLSGen2Deploy.java

```diff
@@ -0,0 +1,252 @@
+
+    @NotNull
+    private List<NameValuePair> createDirReqParams;
+
+    @NotNull
+    private List<NameValuePair> createFileReqParams;
+
+    @NotNull
+    private List<NameValuePair> appendReqParams;
+
+    @NotNull
+    private ADLSGen2ParamsBuilder flushReqParamsBuilder;
+
+    @NotNull
+    private ADLSGen2HeadersBuilder headerBuilder;
+
+    @NotNull
+    private SharedKeyCredentials cred;
+
+    public ADLSGen2Deploy(IClusterDetail cluster, HttpObservable http, String accessKey) {
+        this.cluster = cluster;
+        this.headerBuilder = new ADLSGen2HeadersBuilder();
+
+        try {
+            this.cred = new SharedKeyCredentials(this.cluster.getStorageAccount().getName(), accessKey);
+        } catch (Exception ex) {
+            throw new IllegalArgumentException("Storage account or access key is invalid." + ex.toString());
+        }
+
+        this.createDirReqParams = new ADLSGen2ParamsBuilder()
+                .setResource("directory")
+                .build();
+
+        this.createFileReqParams = new ADLSGen2ParamsBuilder()
+                .setResource("file")
+                .build();
+
+
+        this.appendReqParams = new ADLSGen2ParamsBuilder()
+                .setAction("append")
+                .setPosition(0)
+                .build();
+
+        this.flushReqParamsBuilder = new ADLSGen2ParamsBuilder()
+                .setAction("flush");
+
+        this.http = http;
+        if (http == null) {
+            throw new IllegalArgumentException("Cluster type doesen't support adls gen2 storage type.");
+        }
+
+        http.setDefaultHeaders(headerBuilder.build());
+    }
+
+
+    @Override
+    public URI getUploadDir(@NotNull String rootPath) {
+        return URI.create(rootPath)
+                .resolve(JobUtils.getFormatPathByDate() + "/");
+    }
+
+    @NotNull
+    public Observable<String> deploy(@NotNull File src, @NotNull URI destURI) {
+        //three steps to upload via adls gen2 rest api
+        // 1.put request to create new dir
+        // 2.put request to create new file(artifact) which is empty
+        // 3.patch request to append data to file
+        // 4.patch request to flush data to file
+
+        try {
+            InputStreamEntity reqEntity = new InputStreamEntity(
+                    new FileInputStream(src),
+                    -1,
+                    ContentType.APPLICATION_OCTET_STREAM);
+            BufferedHttpEntity entity = new BufferedHttpEntity(reqEntity);
+            String len = String.valueOf(entity.getContentLength());
+
+            //remove request / end otherwise invalid url response
+            String destStr = destURI.toString();
+            String dirPath = destStr.endsWith("/") ? destStr.substring(0, destStr.length() - 1) : destStr;
+            String filePath = String.format("%s/%s", dirPath, src.getName());
+
+            HttpPut req = new HttpPut(dirPath);
+            return http.request(req, null
+                    , this.createDirReqParams
```

Comment:
Why putting `,` at beginning?

Reply:
first put to create  empty folder then second put to create empty file

---
In reply to: [265846188](https://github.com/Microsoft/azure-tools-for-java/pull/2878#discussion_r265846188) [](ancestors = 265846188)